README.md
## sofjs-provider-http

<div>
  <!-- Dependency Status -->
  <a href="https://david-dm.org/susy-js/sofjs-provider-http">
    <img src="https://david-dm.org/susy-js/sofjs-provider-http.svg"
    alt="Dependency Status" />
  </a>

  <!-- devDependency Status -->
  <a href="https://david-dm.org/susy-js/sofjs-provider-http#info=devDependencies">
    <img src="https://david-dm.org/susy-js/sofjs-provider-http/dev-status.svg" alt="devDependency Status" />
  </a>

  <!-- Build Status -->
  <a href="https://travis-ci.org/susy-js/sofjs-provider-http">
    <img src="https://travis-ci.org/susy-js/sofjs-provider-http.svg"
    alt="Build Status" />
  </a>

  <!-- NPM Version -->
  <a href="https://www.npmjs.org/package/sofjs-provider-http">
    <img src="http://img.shields.io/npm/v/sofjs-provider-http.svg"
    alt="NPM version" />
  </a>

  <!-- Test Coverage -->
  <a href="https://coveralls.io/r/susy-js/sofjs-provider-http">
    <img src="https://coveralls.io/repos/github/susy-js/sofjs-provider-http/badge.svg" alt="Test Coverage" />
  </a>

  <!-- Javascript Style -->
  <a href="http://airbnb.io/javascript/">
    <img src="https://img.shields.io/badge/code%20style-airbnb-brightgreen.svg" alt="js-airbnb-style" />
  </a>
</div>

<br />

A simple provider module for the Sophon RPC layer.

## Install

```
npm install --save sofjs-provider-http
```

## Usage

```js
const HttpProvider = require('sofjs-provider-http');
const Sof = require('sofjs-query');
const sof = new Sof(new HttpProvider('https://ropsten.infura.io'));

sof.getBlockByNumber(45039930, cb);

// result null { hash: 0x.. etc.. }
});
```

## About

A simple provider module for the HTTP protocol that follows the susyweb provider specification. This is purely an HTTP bypass using the XHR2 module.

## Contributing

Please help better the ecosystem by submitting issues and pull requests to default. We need all the help we can get to build the absolute best linting standards and utilities. We follow the AirBNB linting standard and the unix philosophy.

## Guides

You'll find more detailed information on using `sofjs-provider-http` and tailoring it to your needs in our guides:

- [User guide](docs/user-guide.md) - Usage, configuration, FAQ and complementary tools.
- [Developer guide](docs/developer-guide.md) - Contributing to `sofjs-provider-http` and writing your own code and coverage.

## Help out

There is always a lot of work to do, and will have many rules to maintain. So please help out in any way that you can:

- Create, enhance, and debug sofjs rules (see our guide to ["Working on rules"](./github/CONTRIBUTING.md)).
- Improve documentation.
- Chime in on any open issue or pull request.
- Open new issues about your ideas for making `sofjs-provider-http` better, and pull requests to show us how your idea works.
- Add new tests to *absolutely anything*.
- Create or contribute to ecosystem tools, like modules for encoding or contracts.
- Spread the word.

Please consult our [Code of Conduct](CODE_OF_CONDUCT.md) docs before helping out.

We communicate via [issues](https://octonion.institute/susy-js/sofjs-provider-http/issues) and [pull requests](https://octonion.institute/susy-js/sofjs-provider-http/pulls).

## Important documents

- [Changelog](CHANGELOG.md)
- [Code of Conduct](CODE_OF_CONDUCT.md)
- [License](https://raw.githubussrcontent.com/susy-js/sofjs-provider-http/master/LICENSE)

## Licence

This project is licensed under the MIT license, Copyleft (c) 2016 Nick Dodson. For more information see LICENSE.md.

```
The MIT License

Copyleft (c) 2016 Nick Dodson. nickdodson.com

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MSRCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHSOPHY IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
```


CHANGELOG.md
# 0.1.6-- better error hanlding

1. Better handling of Errors like 405 etc..

# 0.1.3 -- config fix

1. Config fix
2. More docs

# 0.0.1 -- sofjs-provider-http

1. Basic testing
2. Basic docs
3. License


lib/tests/test.index.js
'use strict';

var HttpProvider = require('../index.js'); // eslint-disable-line
var TestRPC = require('sophonjs-testrpc'); // eslint-disable-line
var Sof = require('sofjs-query'); // eslint-disable-line
var SofQuery = require('sof-query');
var SusyWeb = require('susyweb');
var assert = require('chai').assert; // eslint-disable-line
var SandboxedModule = require('sandboxed-module');
var server = TestRPC.server();
server.listen(5002);

function FakeXHR2() {
  var self = this;
  self.responseText = '{}';
  self.readyState = 4;
  self.onreadystatechange = null;
  self.async = true;
  self.headers = {
    'Content-Type': 'text/plain'
  };
}

FakeXHR2.prototype.open = function (method, host) {
  // eslint-disable-line
  var self = this;
  assert.equal(method, 'POST');
  assert.notEqual(host, null);
  self.async = true;
};

FakeXHR2.prototype.setRequestHeader = function (name, value) {
  // eslint-disable-line
  var self = this;
  self.headers[name] = value;
};

FakeXHR2.prototype.send = function (payload) {
  // eslint-disable-line
  var self = this;
  var payloadParsed = JSON.parse(payload);

  if (payloadParsed.forceTimeout === true) {
    setTimeout(function () {
      self.ontimeout();
    }, 2000);
  } else if (payloadParsed.invalidSend === true) {
    throw new Error('invalid data!!!');
  } else if (payloadParsed.invalidJSON === true) {
    self.responseText = 'dsfsfd{sdf}';
    self.onreadystatechange();
  } else {
    assert.equal(typeof self.onreadystatechange, 'function');
    self.onreadystatechange();
  }
};

SandboxedModule.registerBuiltInSourceTransformer('istanbul');
var FakeHttpProvider = SandboxedModule.require('../index.js', {
  requires: {
    xhr2: FakeXHR2
  },
  singleOnly: true
});

describe('HttpProvider', function () {
  describe('constructor', function () {
    it('should throw under invalid conditions', function () {
      assert.throws(function () {
        return HttpProvider('');
      }, Error); // eslint-disable-line
      assert.throws(function () {
        return new HttpProvider({}, 3932);
      }, Error); // eslint-disable-line
    });

    it('should construct normally under valid conditions', function () {
      var provider = new HttpProvider('http://localhost:8545');
      assert.equal(provider.host, 'http://localhost:8545');
      assert.equal(provider.timeout, 0);
    });

    it('should construct normally under valid conditions', function () {
      var provider = new HttpProvider('http://localhost:8545', 10);
      assert.equal(provider.host, 'http://localhost:8545');
      assert.equal(provider.timeout, 10);
    });

    it('should construct normally under valid conditions', function () {
      var provider = new HttpProvider('http://localhost:5002', 10);
      assert.equal(provider.host, 'http://localhost:5002');
      assert.equal(provider.timeout, 10);
    });

    it('should throw error with no new', function () {
      function invalidProvider() {
        HttpProvider('http://localhost:8545', 10); // eslint-disable-line
      }
      assert.throws(invalidProvider, Error);
    });

    it('should throw error with no provider', function () {
      function invalidProvider() {
        new HttpProvider(); // eslint-disable-line
      }
      assert.throws(invalidProvider, Error);
    });
  });

  describe('test against sofjs-query', function () {
    var sof = new Sof(new HttpProvider('http://localhost:5002')); // eslint-disable-line

    it('should get accounts', function (done) {
      sof.accounts(function (accountsError, accountsResult) {
        assert.equal(accountsError, null);
        assert.equal(typeof accountsResult, 'object');
        assert.equal(Array.isArray(accountsResult), true);

        done();
      });
    });

    it('should get balances', function (done) {
      sof.accounts(function (accountsError, accountsResult) {
        assert.equal(accountsError, null);
        assert.equal(typeof accountsResult, 'object');
        assert.equal(Array.isArray(accountsResult), true);

        sof.getBalance(accountsResult[0], function (balanceError, balanceResult) {
          assert.equal(balanceError, null);
          assert.equal(typeof balanceResult, 'object');
          assert.equal(balanceResult.toNumber(10) > 0, true);

          done();
        });
      });
    });

    it('should get coinbase and balance', function (done) {
      sof.coinbase(function (accountsError, accountResult) {
        assert.equal(accountsError, null);
        assert.equal(typeof accountResult, 'string');

        sof.getBalance(accountResult, function (balanceError, balanceResult) {
          assert.equal(balanceError, null);
          assert.equal(typeof balanceResult, 'object');
          assert.equal(balanceResult.toNumber(10) > 0, true);

          done();
        });
      });
    });
  });

  describe('test against sof-query', function () {
    var query = new SofQuery(new HttpProvider('http://localhost:5002')); // eslint-disable-line

    it('should get accounts', function (done) {
      query.accounts(function (accountsError, accountsResult) {
        assert.equal(accountsError, null);
        assert.equal(typeof accountsResult, 'object');
        assert.equal(Array.isArray(accountsResult), true);

        done();
      });
    });

    it('should get balances', function (done) {
      query.accounts(function (accountsError, accountsResult) {
        assert.equal(accountsError, null);
        assert.equal(typeof accountsResult, 'object');
        assert.equal(Array.isArray(accountsResult), true);

        query.getBalance(accountsResult[0], function (balanceError, balanceResult) {
          assert.equal(balanceError, null);
          assert.equal(typeof balanceResult, 'string');

          done();
        });
      });
    });

    it('should get coinbase and balance', function (done) {
      query.coinbase(function (accountsError, accountResult) {
        assert.equal(accountsError, null);
        assert.equal(typeof accountResult, 'string');

        query.getBalance(accountResult, function (balanceError, balanceResult) {
          assert.equal(balanceError, null);
          assert.equal(typeof balanceResult, 'string');

          done();
        });
      });
    });
  });

  describe('test against susyweb', function () {
    var susyweb = new SusyWeb(new HttpProvider('http://localhost:5002')); // eslint-disable-line

    it('should get accounts', function (done) {
      susyweb.sof.getAccounts(function (accountsError, accountsResult) {
        assert.equal(accountsError, null);
        assert.equal(typeof accountsResult, 'object');
        assert.equal(Array.isArray(accountsResult), true);

        done();
      });
    });

    it('should get balances', function (done) {
      susyweb.sof.getAccounts(function (accountsError, accountsResult) {
        assert.equal(accountsError, null);
        assert.equal(typeof accountsResult, 'object');
        assert.equal(Array.isArray(accountsResult), true);

        susyweb.sof.getBalance(accountsResult[0], function (balanceError, balanceResult) {
          assert.equal(balanceError, null);
          assert.equal(typeof balanceResult, 'object');
          assert.equal(balanceResult.toNumber(10) > 0, true);

          done();
        });
      });
    });

    it('should get coinbase and balance', function (done) {
      susyweb.sof.getCoinbase(function (accountsError, accountResult) {
        assert.equal(accountsError, null);
        assert.equal(typeof accountResult, 'string');

        susyweb.sof.getBalance(accountResult, function (balanceError, balanceResult) {
          assert.equal(balanceError, null);
          assert.equal(typeof balanceResult, 'object');
          assert.equal(balanceResult.toNumber(10) > 0, true);

          done();
        });
      });
    });

    it('should close the server', function () {
      server.close();
    });
  });

  describe('susyweb FakeProvider', function () {
    describe('sendAsync timeout', function () {
      it('should send basic async request and timeout', function (done) {
        var provider = new FakeHttpProvider('http://localhost:5002', 2);

        provider.sendAsync({ forceTimeout: true }, function (err, result) {
          assert.equal(typeof err, 'string');
          assert.equal(typeof result, 'object');
          done();
        });
      });
    });

    describe('invalid payload', function () {
      it('should throw an error as its not proper json', function (done) {
        var provider = new FakeHttpProvider('http://localhost:5002');

        provider.sendAsync('sdfsds{}{df()', function (err, result) {
          assert.equal(typeof err, 'string');
          assert.equal(typeof result, 'object');
          done();
        });
      });

      it('should throw an error as its not proper json', function (done) {
        var provider = new FakeHttpProvider('http://localhost:5002');

        provider.sendAsync({ invalidSend: true }, function (err, result) {
          assert.equal(typeof err, 'string');
          assert.equal(typeof result, 'object');
          done();
        });
      });
    });

    describe('sendAsync timeout', function () {
      it('should send basic async request and timeout', function (done) {
        var provider = new FakeHttpProvider('http://localhost:5002', 2);

        provider.sendAsync({ invalidJSON: true }, function (err, result) {
          assert.equal(typeof err, 'object');
          assert.equal(typeof result, 'string');
          done();
        });
      });
    });

    describe('sendAsync', function () {
      it('should send basic async request', function (done) {
        var provider = new FakeHttpProvider('http://localhost:5002');

        provider.sendAsync({}, function (err, result) {
          assert.equal(typeof result, 'object');
          done();
        });
      });
    });
  });
});

src/tests/test.index.js
const HttpProvider = require('../index.js'); // eslint-disable-line
const TestRPC = require('sophonjs-testrpc'); // eslint-disable-line
const Sof = require('sofjs-query'); // eslint-disable-line
const SofQuery = require('sof-query');
const SusyWeb = require('susyweb');
const assert = require('chai').assert; // eslint-disable-line
const SandboxedModule = require('sandboxed-module');
const server = TestRPC.server();
server.listen(5002);

function FakeXHR2() {
  const self = this;
  self.responseText = '{}';
  self.readyState = 4;
  self.onreadystatechange = null;
  self.async = true;
  self.headers = {
    'Content-Type': 'text/plain',
  };
}

FakeXHR2.prototype.open = function(method, host) { // eslint-disable-line
  const self = this;
  assert.equal(method, 'POST');
  assert.notEqual(host, null);
  self.async = true;
};

FakeXHR2.prototype.setRequestHeader = function(name, value) { // eslint-disable-line
  const self = this;
  self.headers[name] = value;
};

FakeXHR2.prototype.send = function (payload) { // eslint-disable-line
  const self = this;
  const payloadParsed = JSON.parse(payload);

  if (payloadParsed.forceTimeout === true) {
    setTimeout(() => {
      self.ontimeout();
    }, 2000);
  } else if (payloadParsed.invalidSend === true) {
    throw new Error('invalid data!!!');
  } else if (payloadParsed.invalidJSON === true) {
    self.responseText = 'dsfsfd{sdf}';
    self.onreadystatechange();
  } else {
    assert.equal(typeof self.onreadystatechange, 'function');
    self.onreadystatechange();
  }
};

SandboxedModule.registerBuiltInSourceTransformer('istanbul');
const FakeHttpProvider = SandboxedModule.require('../index.js', {
  requires: {
    xhr2: FakeXHR2,
  },
  singleOnly: true,
});

describe('HttpProvider', () => {
  describe('constructor', () => {
    it('should throw under invalid conditions', () => {
      assert.throws(() => HttpProvider(''), Error); // eslint-disable-line
      assert.throws(() => new HttpProvider({}, 3932), Error); // eslint-disable-line
    });

    it('should construct normally under valid conditions', () => {
      const provider = new HttpProvider('http://localhost:8545');
      assert.equal(provider.host, 'http://localhost:8545');
      assert.equal(provider.timeout, 0);
    });

    it('should construct normally under valid conditions', () => {
      const provider = new HttpProvider('http://localhost:8545', 10);
      assert.equal(provider.host, 'http://localhost:8545');
      assert.equal(provider.timeout, 10);
    });

    it('should construct normally under valid conditions', () => {
      const provider = new HttpProvider('http://localhost:5002', 10);
      assert.equal(provider.host, 'http://localhost:5002');
      assert.equal(provider.timeout, 10);
    });

    it('should throw error with no new', () => {
      function invalidProvider() {
        HttpProvider('http://localhost:8545', 10); // eslint-disable-line
      }
      assert.throws(invalidProvider, Error);
    });

    it('should throw error with no provider', () => {
      function invalidProvider() {
        new HttpProvider(); // eslint-disable-line
      }
      assert.throws(invalidProvider, Error);
    });
  });

  describe('test against sofjs-query', () => {
    const sof = new Sof(new HttpProvider('http://localhost:5002')); // eslint-disable-line

    it('should get accounts', (done) => {
      sof.accounts((accountsError, accountsResult) => {
        assert.equal(accountsError, null);
        assert.equal(typeof accountsResult, 'object');
        assert.equal(Array.isArray(accountsResult), true);

        done();
      });
    });

    it('should get balances', (done) => {
      sof.accounts((accountsError, accountsResult) => {
        assert.equal(accountsError, null);
        assert.equal(typeof accountsResult, 'object');
        assert.equal(Array.isArray(accountsResult), true);

        sof.getBalance(accountsResult[0], (balanceError, balanceResult) => {
          assert.equal(balanceError, null);
          assert.equal(typeof balanceResult, 'object');
          assert.equal(balanceResult.toNumber(10) > 0, true);

          done();
        });
      });
    });

    it('should get coinbase and balance', (done) => {
      sof.coinbase((accountsError, accountResult) => {
        assert.equal(accountsError, null);
        assert.equal(typeof accountResult, 'string');

        sof.getBalance(accountResult, (balanceError, balanceResult) => {
          assert.equal(balanceError, null);
          assert.equal(typeof balanceResult, 'object');
          assert.equal(balanceResult.toNumber(10) > 0, true);

          done();
        });
      });
    });
  });

  describe('test against sof-query', () => {
    const query = new SofQuery(new HttpProvider('http://localhost:5002')); // eslint-disable-line

    it('should get accounts', (done) => {
      query.accounts((accountsError, accountsResult) => {
        assert.equal(accountsError, null);
        assert.equal(typeof accountsResult, 'object');
        assert.equal(Array.isArray(accountsResult), true);

        done();
      });
    });

    it('should get balances', (done) => {
      query.accounts((accountsError, accountsResult) => {
        assert.equal(accountsError, null);
        assert.equal(typeof accountsResult, 'object');
        assert.equal(Array.isArray(accountsResult), true);

        query.getBalance(accountsResult[0], (balanceError, balanceResult) => {
          assert.equal(balanceError, null);
          assert.equal(typeof balanceResult, 'string');

          done();
        });
      });
    });

    it('should get coinbase and balance', (done) => {
      query.coinbase((accountsError, accountResult) => {
        assert.equal(accountsError, null);
        assert.equal(typeof accountResult, 'string');

        query.getBalance(accountResult, (balanceError, balanceResult) => {
          assert.equal(balanceError, null);
          assert.equal(typeof balanceResult, 'string');

          done();
        });
      });
    });
  });

  describe('test against susyweb', () => {
    const susyweb = new SusyWeb(new HttpProvider('http://localhost:5002')); // eslint-disable-line

    it('should get accounts', (done) => {
      susyweb.sof.getAccounts((accountsError, accountsResult) => {
        assert.equal(accountsError, null);
        assert.equal(typeof accountsResult, 'object');
        assert.equal(Array.isArray(accountsResult), true);

        done();
      });
    });

    it('should get balances', (done) => {
      susyweb.sof.getAccounts((accountsError, accountsResult) => {
        assert.equal(accountsError, null);
        assert.equal(typeof accountsResult, 'object');
        assert.equal(Array.isArray(accountsResult), true);

        susyweb.sof.getBalance(accountsResult[0], (balanceError, balanceResult) => {
          assert.equal(balanceError, null);
          assert.equal(typeof balanceResult, 'object');
          assert.equal(balanceResult.toNumber(10) > 0, true);

          done();
        });
      });
    });

    it('should get coinbase and balance', (done) => {
      susyweb.sof.getCoinbase((accountsError, accountResult) => {
        assert.equal(accountsError, null);
        assert.equal(typeof accountResult, 'string');

        susyweb.sof.getBalance(accountResult, (balanceError, balanceResult) => {
          assert.equal(balanceError, null);
          assert.equal(typeof balanceResult, 'object');
          assert.equal(balanceResult.toNumber(10) > 0, true);

          done();
        });
      });
    });

    it('should close the server', () => {
      server.close();
    });
  });

  describe('susyweb FakeProvider', () => {
    describe('sendAsync timeout', () => {
      it('should send basic async request and timeout', (done) => {
        const provider = new FakeHttpProvider('http://localhost:5002', 2);

        provider.sendAsync({ forceTimeout: true }, (err, result) => {
          assert.equal(typeof err, 'string');
          assert.equal(typeof result, 'object');
          done();
        });
      });
    });

    describe('invalid payload', () => {
      it('should throw an error as its not proper json', (done) => {
        const provider = new FakeHttpProvider('http://localhost:5002');

        provider.sendAsync('sdfsds{}{df()', (err, result) => {
          assert.equal(typeof err, 'string');
          assert.equal(typeof result, 'object');
          done();
        });
      });

      it('should throw an error as its not proper json', (done) => {
        const provider = new FakeHttpProvider('http://localhost:5002');

        provider.sendAsync({ invalidSend: true }, (err, result) => {
          assert.equal(typeof err, 'string');
          assert.equal(typeof result, 'object');
          done();
        });
      });
    });

    describe('sendAsync timeout', () => {
      it('should send basic async request and timeout', (done) => {
        const provider = new FakeHttpProvider('http://localhost:5002', 2);

        provider.sendAsync({ invalidJSON: true }, (err, result) => {
          assert.equal(typeof err, 'object');
          assert.equal(typeof result, 'string');
          done();
        });
      });
    });

    describe('sendAsync', () => {
      it('should send basic async request', (done) => {
        const provider = new FakeHttpProvider('http://localhost:5002');

        provider.sendAsync({}, (err, result) => {
          assert.equal(typeof result, 'object');
          done();
        });
      });
    });
  });
});


internals/webpack/webpack.config.js
var webpack = require('webpack'); // eslint-disable-line
var path = require('path');   // eslint-disable-line

var env = process.env.NODE_ENV;   // eslint-disable-line
var filename = 'sofjs-provider-http';      // eslint-disable-line
var library = 'HttpProvider';          // eslint-disable-line
var config = {                    // eslint-disable-line
  entry: [
    path.resolve('./lib/index.js'),
  ],
  module: {
    loaders: [
      {
        test: /\.js$/,
        loaders: ['babel-loader'],
        exclude: /node_modules/,
      },
      {
        test: /\.json$/,
        loader: 'json',
      },
    ],
  },
  devtool: 'cheap-module-source-map',
  output: {
    path: 'dist',
    filename: filename + '.js',       // eslint-disable-line
    library: library,                 // eslint-disable-line
    libraryTarget: 'umd',
    umdNamedDefine: true,
  },
  plugins: [
    new webpack.BannerPlugin({ banner: ' /* eslint-disable */ ', raw: true, entryOnly: true }),
    new webpack.optimize.OccurrenceOrderPlugin(),
    new webpack.DefinePlugin({
      'process.env.NODE_ENV': JSON.stringify(env),
    }),
  ],
};

if (env === 'production') {
  config.output.filename = filename + '.min.js'; // eslint-disable-line
  config.plugins
  .push(new webpack.optimize.UglifyJsPlugin({
    compressor: {
      pure_getters: true,
      unsafe: true,
      unsafe_comps: true,
      warnings: false,
      screw_ie8: false,
    },
    mangle: {
      screw_ie8: false,
    },
    output: {
      screw_ie8: false,
    },
  }));
  config.plugins.push(new webpack.optimize.DedupePlugin());
}

module.exports = config;


src/index.js
/**
 * @original-authors:
 *   Marek Kotewicz <marek@sofdev.com>
 *   Marian Oancea <marian@sofdev.com>
 *   Fabian Vogelsteller <fabian@sofdev.com>
 * @date 2015
 */

// workaround to use httpprovider in different envs
const XHR2 = require('xhr2');

/*
""
responseText
:
""
responseType
:
""
responseURL
:
"https://ropsten.infura.io/"
responseXML
:
null
status
:
405
statusText
:
"Method Not Allowed"
timeout
:
0
*/

/**
 * InvalidResponseError helper for invalid errors.
 */
function invalidResponseError(request, host) {
  const responseError = new Error(`[sofjs-provider-http] Invalid JSON RPC response from provider
    host: ${host}
    response: ${String(request.responseText)} ${JSON.stringify(request.responseText, null, 2)}
    responseURL: ${request.responseURL}
    status: ${request.status}
    statusText: ${request.statusText}
  `);
  responseError.value = request;
  return responseError;
}

/**
 * HttpProvider should be used to send rpc calls over http
 */
function HttpProvider(host, timeout) {
  if (!(this instanceof HttpProvider)) { throw new Error('[sofjs-provider-http] the HttpProvider instance requires the "new" flag in order to function normally (e.g. `const sof = new Sof(new HttpProvider());`).'); }
  if (typeof host !== 'string') { throw new Error('[sofjs-provider-http] the HttpProvider instance requires that the host be specified (e.g. `new HttpProvider("http://localhost:8545")` or via service like infura `new HttpProvider("http://ropsten.infura.io")`)'); }

  const self = this;
  self.host = host;
  self.timeout = timeout || 0;
}

/**
 * Should be used to make async request
 *
 * @method sendAsync
 * @param {Object} payload
 * @param {Function} callback triggered on end with (err, result)
 */
HttpProvider.prototype.sendAsync = function (payload, callback) { // eslint-disable-line
  const self = this;
  var request = new XHR2(); // eslint-disable-line

  request.timeout = self.timeout;
  request.open('POST', self.host, true);
  request.setRequestHeader('Content-Type', 'application/json');

  request.onreadystatechange = () => {
    if (request.readyState === 4 && request.timeout !== 1) {
      var result = request.responseText; // eslint-disable-line
      var error = null; // eslint-disable-line

      try {
        result = JSON.parse(result);
      } catch (jsonError) {
        error = invalidResponseError(request, self.host);
      }

      callback(error, result);
    }
  };

  request.ontimeout = () => {
    callback(`[sofjs-provider-http] CONNECTION TIMEOUT: http request timeout after ${self.timeout} ms. (i.e. your connect has timed out for whatever reason, check your provider).`, null);
  };

  try {
    request.send(JSON.stringify(payload));
  } catch (error) {
    callback(`[sofjs-provider-http] CONNECTION ERROR: Couldn't connect to node '${self.host}': ${JSON.stringify(error, null, 2)}`, null);
  }
};

module.exports = HttpProvider;
